File: aitools/proofs/builtin_provers.py
import logging

from aitools.logic.core import Expression, Variable, LogicObject
from aitools.logic.unification import Substitution
from aitools.logic.utils import VariableSource
from aitools.proofs.components import HandlerArgumentMode, HandlerSafety
from aitools.proofs.knowledge_base import KnowledgeBase
from aitools.proofs.language import Implies, Not
from aitools.proofs.provers import TruthSubstitutionPremises, Prover, TruthSubstitution

logger = logging.getLogger(__name__)


async def restricted_modus_ponens(formula: LogicObject, substitution: Substitution, kb: KnowledgeBase):
    """Restricted backward version of modus ponens, which won't perform recursive proof of implications"""
    v = VariableSource()
    if not (isinstance(formula, Expression) and formula.children[0] == Implies):
        rule_pattern = Implies(v.premise, formula)

        async for rule_proof in kb.async_prove(rule_pattern, previous_substitution=substitution):
            premise = rule_proof.substitution.get_bound_object_for(v.premise)
            async for premise_proof in kb.async_prove(premise, previous_substitution=rule_proof.substitution):
                yield TruthSubstitutionPremises(truth=True, substitution=premise_proof.substitution,
                                                premises=(rule_proof, premise_proof))


RestrictedModusPonens = Prover(
    listened_formula=Variable(), handler=restricted_modus_ponens, argument_mode=HandlerArgumentMode.RAW,
    pass_substitution_as=..., pass_knowledge_base_as='kb', pure=True, safety=HandlerSafety.SAFE
)


async def closed_world_assumption(formula: LogicObject, substitution: Substitution, kb: KnowledgeBase):
    v = VariableSource()
    match = Substitution.unify(formula, Not(v.P))
    if match is not None:
        try:
            await kb.async_prove(match.get_bound_object_for(v.P)).__anext__()
        except StopAsyncIteration:
            return TruthSubstitution(True, substitution)


ClosedWorldAssumption = Prover(
    listened_formula=Variable(), handler=closed_world_assumption, argument_mode=HandlerArgumentMode.RAW,
    pass_substitution_as=..., pass_knowledge_base_as='kb', pure=True, safety=HandlerSafety.SAFE
)

File: AgainIsThisAGoodIdea/Proofs/BuiltinProvers.fs
module AITools.Proofs.Builtin.Provers

open AITools.Logic.Language
open AITools.Logic.Utils
open AITools.Logic.Core
open AITools.Proofs.Components.Base
open AITools.Proofs.Components.Provers
open AITools.Proofs.KnowledgeBase
open AITools.Proofs.Language
open System.Collections.Immutable
open AITools.Utils.AsyncTools

let private language = Language()

// TODO this can't be a constant here, but still we need it somewhere
let bufferSize = 1

let private restrictedModusPonens (input:{|expression: _; substitution:_; kb: KnowledgeBase|}) return' = async{
    let v = VariableSource language
    
    match input.expression with
    | Expr children when children.Length > 0 && children.[0] = Implies -> ()
    | _ ->
        let rulePattern = [|Implies; Var(v?premise); input.expression|] |> ImmutableArray.CreateRange |> Expr
        let ruleProofs = input.kb.AsyncProve (rulePattern, false, input.substitution, bufferSize)
        do! foreachResultParallel bufferSize ruleProofs <| fun ruleProof -> async{
            match ruleProof.Substitution.GetBoundObjectFor(v?premise) with
            | Some premise -> 
                let premiseProofs = input.kb.AsyncProve(premise, false, ruleProof.Substitution, bufferSize)
                do! foreachResultParallel bufferSize premiseProofs <| fun premiseProof ->
                    return' (true, premiseProof.Substitution, seq{ruleProof; premiseProof})
            | None -> ()
        }
}


let RestrictedModusPonens: Prover<KnowledgeBase> = 
    makeProver <| HandlerDescriptor.MakeRaw(
        make language VarExpr,
        AsyncSourcePremisedSatisfier restrictedModusPonens,
        HandlerPurity.Impure,
        HandlerSafety.Safe,
        PassSubstitutionAs "substitution",
        PassContextAs "kb"
    )


// async def closed_world_assumption(formula: LogicObject, substitution: Substitution, kb: KnowledgeBase):
//     language = Language()
//     v = VariableSource(language=language)
//     match = Substitution.unify(formula, Not(v.P))
//     if match is not None:
//         try:
//             await kb.async_prove(match.get_bound_object_for(v.P)).__anext__()
//         except StopAsyncIteration:
//             return TruthSubstitution(True, substitution)


// ClosedWorldAssumption = Prover(
//     listened_formula=Variable(language=language), handler=closed_world_assumption, argument_mode=HandlerArgumentMode.RAW,
//     pass_substitution_as=..., pass_knowledge_base_as='kb', pure=True, safety=HandlerSafety.SAFE
// )